Mathlib/Algebra/Module/CharacterModule.lean

```diff
@@ -0,0 +1,143 @@
+/-
+Copyright (c) 2023 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang, Junyan Xu
+-/
+
+import Mathlib.Algebra.Module.LinearMap
+import Mathlib.Algebra.Category.ModuleCat.Basic
+import Mathlib.Algebra.Category.GroupCat.Injective
+import Mathlib.Topology.Instances.AddCircle
+import Mathlib.Topology.Instances.Rat
+import Mathlib.Algebra.Category.ModuleCat.ChangeOfRings
+
+/-!
+# Character module of a module
+
+For commutative ring `R` and an `R`-module `M` and an injective module `D`, its character module
+`M⋆` is defined to be `R`-linear maps `M ⟶ D`.
+
+`M⋆` also has an `R`-module structure given by `(r • f) m = f (r • m)`.
+
+## Main results
+
+- `CharacterModuleFunctor` : the contravariant functor of `R`-modules where `M ↦ M⋆` and
+an `R`-lineara map `l : M ⟶ N` induces an `R`-linear map `l⋆ : f ↦ f ∘ l` where `f : N⋆`.
+- `LinearMap.charaterfy_surjective_of_injective` : If `l` is injective then `l⋆` is surjective,
+  in another word taking character module as a functor sends monos to epis.
+- `CharacterModule.exists_character_apply_ne_zero_of_ne_zero` : for nonzero `a ∈ M`, there is a
+  character `c` in `M⋆` such that `c a` is nonzero as well.
+- `CharacterModule.homEquiv` : there is a bijection between linear map `Hom(N, M⋆)` and
+  `(N ⊗ M)⋆` given by `curry` and `uncurry`.
+
+-/
+
+open CategoryTheory
+
+universe uR uM uN uD
+variable (R : Type uR) [CommRing R]
+variable (M : Type (max uR uM)) [AddCommGroup M] [Module R M]
+variable (N : Type (max uR uN)) [AddCommGroup N] [Module R N]
+variable (D : Type (max uR uD)) [AddCommGroup D] [Module R D]
+variable [injective_dual : Injective <| ModuleCat.of R D]
```

Comment:
In principle, the Lean compiler should not use this hypothesis that D is injective unless it is needed, but can you make sure that this is the case?

Reply:
I think @AntoineChambert-Loir means that you should only include the injectivity hypothesis when needed, rather than having it as a `variable` (so that readers of the source code (e.g. PR reviewers) can't easily figure out whether injectivity is used or not (even though it's easy to figure out in VSCode or the online docs)).

Reply:
I suggest that you make this fully universe polymorphic and specialize only when necessary, and replace `CategoryTheory.Injective` by `Module.Injective` here.

Reply:
This is done, I will see if flatness still works using `R`-linear map later in another branch